tools/dependency/validate.py

```diff
@@ -0,0 +1,260 @@
+_repository_locations_spec.loader.exec_module(repository_locations)
+
+# source/extensions/extensions_build_config.bzl must have a .bzl suffix for Starlark
+# import, so we are forced to do this workaround.
+_extensions_build_config_spec = spec_from_loader(
+    'extensions_build_config',
+    SourceFileLoader('extensions_build_config', 'source/extensions/extensions_build_config.bzl'))
+extensions_build_config = module_from_spec(_extensions_build_config_spec)
+_extensions_build_config_spec.loader.exec_module(extensions_build_config)
+
+BAZEL_QUERY_EXTERNAL_DEP_RE = re.compile('@(\w+)//')
+EXTENSION_LABEL_RE = re.compile('(//source/extensions/.*):')
+
+# TODO(htuch): Add API dependencies to metadata, shrink this set.
+UNKNOWN_DEPS = [
+    'org_golang_x_tools', 'com_github_cncf_udpa', 'org_golang_google_protobuf',
+    'io_bazel_rules_nogo', 'com_envoyproxy_protoc_gen_validate', 'opencensus_proto',
+    'io_bazel_rules_go', 'foreign_cc_platform_utils', 'com_github_golang_protobuf',
+    'com_google_googleapis'
+]
+IGNORE_DEPS = set(['envoy', 'envoy_api', 'platforms', 'bazel_tools', 'local_config_cc'] +
+                  UNKNOWN_DEPS)
+
+
+class DependencyError(Exception):
+  """Error in dependency relationships."""
+  pass
+
+
+class DependencyInfo(object):
+  """Models dependency info in bazel/repositories.bzl."""
+
+  def DepsByUseCategory(self, use_category):
+    """Find the set of external dependencies in a given use_category.
+
+    Args:
+      use_category: string providing use_category.
+
+    Returns:
+      Set of dependency identifiers that match use_category.
+    """
+    deps = set([])
+    for name, metadata in repository_locations.DEPENDENCY_REPOSITORIES.items():
+      if use_category in metadata['use_category']:
+        deps.add(name)
+    return deps
+
+  def GetMetadata(self, dependency):
+    """Obtain repository metadata for a dependency.
+
+    Args:
+      dependency: string providing dependency identifier.
+
+    Returns:
+      A dictionary with the repository metadata as defined in
+      bazel/repository_locations.bzl.
+    """
+    return repository_locations.DEPENDENCY_REPOSITORIES.get(dependency, None)
+
+
+class BuildGraph(object):
+  """Models the Bazel build graph."""
+
+  def QueryExternalDeps(self, *targets):
+    """Query the build graph for transitive external dependencies.
+
+    Args:
+      targets: Bazel targets.
+
+    Returns:
+      A set of dependency identifiers that are reachable from targets.
+    """
+    deps_query = ' union '.join(f'deps({l})' for l in targets)
+    deps = subprocess.check_output(['bazel', 'query', deps_query],
+                                   stderr=subprocess.PIPE).decode().splitlines()
+    ext_deps = set([])
+    for d in deps:
+      match = BAZEL_QUERY_EXTERNAL_DEP_RE.match(d)
+      if match:
+        ext_dep = match.group(1)
+        if ext_dep not in IGNORE_DEPS:
+          ext_deps.add(ext_dep)
+    return set(ext_deps)
```

Comment:
i think this
```
  return set(
    BAZEL_QUERY_EXTERNAL_DEP_RE.match(d).group(1)
    for d
    in deps
    if (BAZEL_QUERY_EXTERNAL_DEP_RE.match(d)
	and (BAZEL_QUERY_EXTERNAL_DEP_RE.match(d).group(1)
             not in IGNORE_DEPS)))
```
is arguably easier to grok

Reply:
I also prefer this style, but I've also had feedback from Pythonistas that they prefer imperative forms when comprehensions get large, Python isn't Haskell. I'll use this review as an excuse to do it with a set comprehension.

Reply:
Actually, I take that back, I think I'll leave it as is since we don't have intermediate binding of the match.